CPU LU factorization and linear solvers

abeleinin opened the pull request:
## Proposed changes

Adds `mx.linalg.solve` backend following the [numpy implementation](https://numpy.org/doc/2.0/reference/generated/numpy.linalg.solve.html). 

For the LAPACK routines, I chose to use a combination of `sgetrf` and `sgetrs`, which is equivalent to calling `sgesv` (https://stackoverflow.com/a/36065619) because it removes a transpose of `a` in the C++.

## Checklist

Put an `x` in the boxes that apply.

- [x] I have read the [CONTRIBUTING](https://github.com/ml-explore/mlx/blob/main/CONTRIBUTING.md) document
- [x] I have run `pre-commit run --all-files` to format my code / installed pre-commit prior to committing changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have updated the necessary documentation (if needed)

awni commented:
This looks good to me. I'm debating with myself if we should try to do this in high-level ops rather than add a new primitive for it.

On the one-hand it is definitely going to be way less efficient. On the other hand it will run on the GPU and have transform support out of the box.

I'm also ok landing this and rearranging the back-end at a later time if ti makes sense.

abeleinin commented:
Were you thinking of an approach using `linalg.lu_factor` and `linalg.solve_triangular` instead? Then `linalg.solve` could just use those ops. I think @barronalex mentioned possibly working on `linalg.solve_triangular` in a previous review, so maybe those linalg ops should be a higher priority than just `linalg.solve` support.

awni commented:
> Then linalg.solve could just use those ops. I think @barronalex mentioned possibly working on linalg.solve_triangular

I think exactly that would be better and more flexible than what we have now. I was even wondering if we could do `linalg.solve_triangular` using existing MLX ops rather than wrap the lapack routine.

abeleinin commented:
Ok, working on adding a `LUF` primitive and subsequent `linalg.lu_factor`, `linalg.solve_triangular`, and `linalg.solve` functions.

abeleinin commented:
Added `LUF` primitive which adds support for `linalg.lu` and `linalg.lu_factor` functions. 

Also added two linear solvers, `linalg.solve_triangular` and `linalg.solve` using higher-level ops.

This pr closes #30

sck-at-ucy commented:
This will be really useful, thank you.

acsweet commented:
I was wondering how close this is to being complete?
It would be great for the Keras backend.

awni commented:
We should be able to land it soon.

awni reviewed (approved):
Thanks for the addition!
